File: MATLAB/example1.m
%% Koopman Operator Tutorial -- Duffing Oscillator No. 1 Example
% 
% S. Servadio, D. Arnas, and R. Linares, “A Koopman Operator Tutorial with 
% Othogonal Polynomials.” arXiv, Jul. 14, 2022. doi: 10.48550/arXiv.2111.07485.

clear; close all; clc;
set(groot, 'defaulttextinterpreter','latex');  
set(groot, 'defaultAxesTickLabelInterpreter','latex');  
set(groot, 'defaultLegendInterpreter','latex');
addpath("src\")

%% Equation parameters
sm = 1; % Mass
sk = 1; % Spring constant
sa = 1; % Unit transformation constant
se = 0.001; % Small parameter

%% Other settings
nt = 100;

q0 = 1; % Initial q
p0 = 0.0; % Initial p
x0 = [q0; p0];

tf = 10; % Final time
tk = linspace(0,tf,nt);

%% Differential equation
% Elements in the array f(a,b,c):
% a: dimension where the derivative is performed
% b: term in the derivative
% c = 1: coefficient of the terms in xdot = f(x)
% c != 1: variable 'c' and its exponent in the polynomial term
Nx = 2;  % number of variables
Nt = 2;  % maximum number of terms in f(x)
fx = zeros(Nx,Nt,Nx+1);

% Linear system
fx(1,1,1) = 1; %x2/sm
fx(1,1,3) = 1;
fx(2,1,1) = -sk; %-x1*sk
fx(2,1,2) = 1;

% Perturbing term
%-sk*se*saˆ2*x1ˆ3
fx(2,2,1) = -sk*se*sa^2;
fx(2,2,2) = 3;

%% Koopman Solution
c = 3;
KSol = KoopOpSol(c,fx,x0,tk);
q = KSol(1,:);
p = KSol(2,:);

%% ODE45 results for comparison
param.sm = sm;
param.sk = sk;
param.se = se;
param.sa = sa;
[t,states] = ode45(@(t,x) duffing1(t,x,param), tk, x0);


%% Plotting results
fig1 = figure(1);
scatter(q,p,DisplayName="Koopman")
hold on; 
scatter(states(:,1),states(:,2),"+",DisplayName="analytical")
legend;
hold off

fig2 = figure(2);
plot(tk,q,lineWidth=2,DisplayName="$q_{k}$")
hold on
plot(tk,p,lineWidth=2,DisplayName="$p_{k}$")
plot(tk,states(:,1),'--',lineWidth=2,DisplayName="$q$")
plot(tk,states(:,2),'--',lineWidth=2,DisplayName="$p$")
legend;
hold off


%% Function
function dxdt = duffing1(t,x,p)
    dxdt = zeros(2,1);
    x1 = x(1);
    x2 = x(2);
    dxdt(1) = x2/p.sm;
    dxdt(2) = -p.sk*x1 - p.sk*p.se*p.sa^2*x1^3;
end

File: scripts/example1.jl
## Example 1: Duffing Oscillator with Perturbation

using DifferentialEquations
using KoopOpGalerkin
using GLMakie
# Makie.inline!(true)  # uncomment to show in navigation pane

const KOG = KoopOpGalerkin
const NORMALIZE = false
const REGULARIZE = true

## Equation parameters
sm = 1.0  # Mass
sk = 0.3  # Spring constant
sa = 1.0  # Unit transformation constant
se = 0.001 # Small parameter

## Other settings
nt = 10000
q0 = 1.0  # Initial q
p0 = 0.0  # Initial p
x0 = [q0, p0]
tf = 30.0  # Final time
tk = range(0, tf, length = nt)

## Run ODE45 system to obtain the L∞ norm of x1 and x2
if NORMALIZE
    function duffing1!(dxdt, x, p, t)
        sm, sk, se, sa = p
        x1 = x[1]
        x2 = x[2]
        dxdt[1] = x2 / sm
        dxdt[2] = -sk * x1 - sk * se * sa^2 * x1^3
    end
    p_vector = [sm, sk, se, sa]
    prob = ODEProblem(duffing1!, x0, (0.0, tf), p_vector)
    sol = solve(prob, Tsit5(), saveat = tk)
    x1max = maximum(abs.(sol[1, :]))
    x2max = maximum(abs.(sol[2, :]))
    x0 = x0 ./ [x1max, x2max]
else
    x1max = 1
    x2max = 1
end

## Differential equation
Nx = 2  # Number of variables
Nt = 2  # Maximum number of terms in f(x)
fx = zeros(Float64, Nx, Nt, Nx + 1)

# Linear system
fx[1, 1, 1] = 1.0
fx[1, 1, 3] = 1.0
fx[2, 1, 1] = -sk
fx[2, 1, 2] = 1.0

# Perturbing term
fx[2, 2, 1] = -sk * se * sa^2
fx[2, 2, 2] = 3.0

# Normalize
fx[1, :, 1] ./= x1max
fx[2, :, 1] ./= x2max

## Koopman Solution
c = 3
KSol = KOG.KoopmanSolve(c, fx, x0, tk)
q = KSol[1, :]
p = KSol[2, :]

## Normalized system: ODE45 results for comparison
function duffing1!(dxdt, x, p, t)
    sm, sk, se, sa, x1max, x2max = p
    x1 = x[1]
    x2 = x[2]
    dxdt[1] = (x2 / sm) / x1max
    dxdt[2] = (-sk * x1 - sk * se * sa^2 * x1^3) / x2max
end
p_vector = [sm, sk, se, sa, x1max, x2max]
prob = ODEProblem(duffing1!, x0, (0.0, tf), p_vector)
sol = solve(prob, Tsit5(), saveat = tk)

## EDMD for comparison
# TODO: The EDMD results are incredibly off. Probably something wrong with the implementation.
# dsol = sol(tk, Val{1})  # generate derivative data
dsol = sol(range(tk.step.hi, tf+tk.step.hi, length=nt))

# Normalize the states and derivatives
lb_states = [minimum(sol[i,:]) for i = 1:Nx]
ub_states = [maximum(sol[i,:]) for i = 1:Nx]
states_n = KOG.normalize(sol[:,:], lb_states, ub_states, -1, 1)

lb_dstates = [minimum(dsol[i,:]) for i = 1:Nx]
ub_dstates = [maximum(dsol[i,:]) for i = 1:Nx]
dstates_n = KOG.normalize(dsol[:,:], lb_dstates, ub_dstates, -1, 1)

## Lift the states and derivatives
n_KO = 20
Xm_lift, Xp_lift, Ms = KOG.lift_data_RBF(states_n, dstates_n, n_KO; verbose=true)
if REGULARIZE
    Ã, ro = KOG.TREDMD(Xm_lift, Xp_lift, 10, 5)
else
    Ã, ro = KOG.EDMD(Xm_lift, Xp_lift, 0)
end

## Simulate the EDMD linearized system
function duffing1_EDMD!(dxdt, x, Ã, t)
    dxdt .= Ã * x
end
x0_lift = KOG.lift_state_RBF(x0, n_KO, Ms)
prob = ODEProblem(duffing1_EDMD!, x0_lift[1:ro], (0.0, tf), Ã)
sol_EDMD = solve(prob, Tsit5(), saveat = tk)

## Plotting results
# Create scatter plot for phase space
fig = Figure(size=(800, 600))
ax1 = Axis(fig[1, 1], xlabel = "x1", ylabel="x2")
scatter!(ax1, sol[1, :], sol[2, :], color=:transparent, strokewidth=1, 
        strokecolor=:red, label="Analytical", marker=:circle)
scatter!(ax1, KSol[1, :], KSol[2, :], color=:blue, label="Koopman", marker=:cross)
scatter!(ax1, sol_EDMD[1, :], sol_EDMD[2, :], color=:transparent,
        strokecolor=:green, strokewidth=1, label="EDMD", marker=:diamond)
axislegend(ax1, position=:lt)
ax1.title = "Phase Space"

# Create line plot for time series
ax2 = Axis(fig[2, 1])
lines!(ax2, tk, sol[1, :], color = :red, label = "x1")
lines!(ax2, tk, sol[2, :], color = :purple, label = "x2")
lines!(ax2, tk, KSol[1, :], color = :blue, label = "x1_k", linestyle=:dash)
lines!(ax2, tk, KSol[2, :], color = :green, label = "x2_k", linestyle=:dash)
lines!(ax2, tk, sol_EDMD[1, :], color = :orange, label = "x1_EDMD", linestyle=:dashdot, linewidth=2)
lines!(ax2, tk, sol_EDMD[2, :], color = :pink, label = "x2_EDMD", linestyle=:dashdot, linewidth=2)
axislegend(ax2)
ax2.title = "Time Series"
ax2.xlabel = "t"
ax2.ylabel = "x"

# Display the figure
fig